src/components/shapes/draw.js

```diff
@@ -579,6 +599,208 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
     }
 }
 
+function drawLabel(gd, index, options, shapeGroup) {
+    // Remove existing label
+    shapeGroup.selectAll('.shape-label').remove();
+
+    // If no label, return
+    if(!options.label) return;
+
+    var labelGroupAttrs = {
+        'data-index': index,
+    };
+    var text = options.label.text;
+    var font = options.label.font;
+
+    var labelTextAttrs = {
+        'data-notex': 1
+    };
+
+    var labelGroup = shapeGroup.append('g')
+        .attr(labelGroupAttrs)
+        .classed('shape-label', true);
+    var labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+
+    // If x0, x1, y0, y1 are defined explicitly, use those values
+    // Otherwise, use shape bounding box
+    var shapex0, shapex1, shapey0, shapey1;
+    if('x0' in options && 'x1' in options && 'y0' in options && 'y1' in options) {
+        // Setup conversion functions
+        var xa = Axes.getFromId(gd, options.xref);
+        var xRefType = Axes.getRefType(options.xref);
+        var ya = Axes.getFromId(gd, options.yref);
+        var yRefType = Axes.getRefType(options.yref);
+        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
+        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
+
+        shapex0 = x2p(options.x0);
+        shapex1 = x2p(options.x1);
+        shapey0 = y2p(options.y0);
+        shapey1 = y2p(options.y1);
+    } else {
+        // Get shape bounding box
+        var shapeBBox = shapeGroup.selectAll('path').node().getBoundingClientRect();
+        shapex0 = shapeBBox.x;
+        shapex1 = shapeBBox.right;
+        shapey0 = shapeBBox.y;
+        shapey1 = shapeBBox.bottom;
+    }
```

Comment:
The vertical positioning for paths is not correct.
Please add a new mock like this and commit the baseline after the changes I proposed above for reversed axes.
Thank you!
```json
{
    "data": [
        {
            "x": [
                0,
                50
            ],
            "y": [
                0,
                50
            ]
        }
    ],
    "layout": {
        "width": 800,
        "height": 600,
        "margin": {
            "t": 100,
            "b": 50,
            "l": 100,
            "r": 50
        },
        "yaxis": {
            "autorange": "reversed"
        },
        "template": {
            "layout": {
                "shapes": [
                    {
                        "label": {"text": "label"},
                        "name": "myPath",
                        "editable": true,
                        "layer": "below",
                        "line": {
                            "width": 0
                        },
                        "fillcolor": "gray",
                        "opacity": 0.5,
                        "xref": "paper",
                        "yref": "paper",
                        "path": "M0.5,0.3C0.5,0.9 0.9,0.9 0.9,0.3C0.9,0.1 0.5,0.1 0.5,0.3ZM0.6,0.4C0.6,0.5 0.66,0.5 0.66,0.4ZM0.74,0.4C0.74,0.5 0.8,0.5 0.8,0.4ZM0.6,0.3C0.63,0.2 0.77,0.2 0.8,0.3Z"
                    }
                ]
            }
        },
        "shapes": [
            {
                "label": {"text": "label"},
                "editable": true,
                "layer": "below",
                "type": "rect",
                "line": {
                    "width": 5
                },
                "fillcolor": "red",
                "opacity": 0.5,
                "xref": "xaxis",
                "yref": "yaxis",
                "x0": 25,
                "y0": 25,
                "x1": 75,
                "y1": 75
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "layer": "top",
                "type": "circle",
                "line": {
                    "width": 5
                },
                "fillcolor": "green",
                "opacity": 0.5,
                "xref": "xaxis",
                "yref": "yaxis",
                "x0": 125,
                "y0": 25,
                "x1": 175,
                "y1": 75
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "line": {
                    "width": 5
                },
                "fillcolor": "blue",
                "path": "M250,25L225,75L275,75Z"
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "line": {
                    "width": 15
                },
                "path": "M250,225L225,275L275,275"
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "layer": "below",
                "path": "M320,100C390,180 290,180 360,100Z",
                "fillcolor": "rgba(0,127,127,0.5)",
                "line": {
                    "width": 5
                }
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "line": {
                    "width": 5,
                    "color": "orange"
                },
                "fillcolor": "rgba(127,255,127,0.5)",
                "path": "M0,100V200H50L0,300Q100,300 100,200T150,200C100,300 200,300 200,200S150,200 150,100Z"
            },
            {
                "label": {"text": "label"},
                "editable": true,
                "line": {
                    "width": 2
                },
                "fillcolor": "yellow",
                "path": "M300,70C300,10 380,10 380,70C380,90 300,90 300,70ZM320,60C320,50 332,50 332,60ZM348,60C348,50 360,50 360,60ZM320,70C326,80 354,80 360,70Z"
            }
        ]
    },
    "config": {
        "editable": false,
        "modeBarButtonsToAdd": [
            "drawline",
            "drawopenpath",
            "drawclosedpath",
            "drawcircle",
            "drawrect",
            "eraseshape"
        ]
    }
}
```

Reply:
@archmoj Mock added!

I'm a bit stumped on how to fix the vertical position for paths -- it seems `getBoundingClientRect()` is giving the wrong y-axis bounding box coords. Is there a different function I should be using?

Reply:
> @archmoj Mock added!
> 

@emilykl I don't see the mock. Did you commit & push?

Reply:
@archmoj Whoops, forgot to push -- should be good now!

Reply:
> I'm a bit stumped on how to fix the vertical position for paths -- it seems getBoundingClientRect() is giving the wrong y-axis bounding box coords. Is there a different function I should be using?

Let's rewrite this part as
```js
    var shapex0, shapex1, shapey0, shapey1;
    if(options.path) {
        var d = getPathString(gd, options);
        var polygons = readPaths(d, gd);
        shapex0 = Infinity;
        shapey0 = Infinity;
        shapex1 = -Infinity;
        shapey1 = -Infinity;
        for(var i = 0; i < polygons.length; i++) {
            var polygon = polygons[i];
            for(var j = 0; j < polygon.length; j++) {
                var _x = polygon[j][1];
                var _y = polygon[j][2];

                shapex0 = Math.min(shapex0, _x);
                shapex1 = Math.max(shapex1, _x);
                shapey0 = Math.min(shapey0, _y);
                shapey1 = Math.max(shapey1, _y);
            }
        }
    } else {
        // Setup conversion functions
        var xa = Axes.getFromId(gd, options.xref);
        var xRefType = Axes.getRefType(options.xref);
        var ya = Axes.getFromId(gd, options.yref);
        var yRefType = Axes.getRefType(options.yref);
        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
        shapex0 = x2p(options.x0);
        shapex1 = x2p(options.x1);
        shapey0 = y2p(options.y0);
        shapey1 = y2p(options.y1);
    }
```

Reply:
To take into account of extra vertices for curves, the following would be a better option:
```js
    var shapex0, shapex1, shapey0, shapey1;
    if(options.path) {
        var d = getPathString(gd, options);
        var polygons = readPaths(d, gd);
        shapex0 = Infinity;
        shapey0 = Infinity;
        shapex1 = -Infinity;
        shapey1 = -Infinity;
        for(var i = 0; i < polygons.length; i++) {
            for(var j = 0; j < polygons[i].length; j++) {
                var p = polygons[i][j];
                for(var k = 1; k < p.length; k += 2) {
                    var _x = p[k];
                    var _y = p[k + 1];

                    shapex0 = Math.min(shapex0, _x);
                    shapex1 = Math.max(shapex1, _x);
                    shapey0 = Math.min(shapey0, _y);
                    shapey1 = Math.max(shapey1, _y);
                }
            }
        }
    } else {
        // Setup conversion functions
        var xa = Axes.getFromId(gd, options.xref);
        var xRefType = Axes.getRefType(options.xref);
        var ya = Axes.getFromId(gd, options.yref);
        var yRefType = Axes.getRefType(options.yref);
        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
        shapex0 = x2p(options.x0);
        shapex1 = x2p(options.x1);
        shapey0 = y2p(options.y0);
        shapey1 = y2p(options.y1);
    }
```

Reply:
thanks @archmoj ! Just pushed your suggested code and this works great.

The text position still looks a little off for this one shape though:
![Screen Shot 2023-02-28 at 10 41 13 AM](https://user-images.githubusercontent.com/4672118/221903307-b848a9c1-67de-4b00-a405-1b348d927262.png)

Reply:
Let's skip this part for now & come back to it later.

Reply:
That particular shape is funny, I don't know its path definition but the curve may extend well below its control points just because of the way bezier curves work, so I wouldn't be too worried about where the labels appear there. Not worth jumping through crazy hoops to improve that, it's a pretty esoteric edge case.